network-mux/demo/cardano-ping.hs

```diff
@@ -0,0 +1,341 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE RankNTypes        #-}
+
+module Main (main) where
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Read     as CBOR
+import qualified Codec.CBOR.Write    as CBOR
+import           Control.Exception
+import           Control.Monad (replicateM, when, unless)
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadTime
+import           Control.Monad.Class.MonadTimer hiding (timeout)
+import           Control.Tracer (nullTracer)
+import           Data.Aeson
+import           Data.ByteString.Lazy (ByteString)
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Char8 as C8 (putStrLn)
+import           Data.Maybe (fromMaybe)
+import           Data.Word
+import           Data.TDigest
+import           Data.Text (unpack)
+import           Network.Socket ( AddrInfo)
+import qualified Network.Socket as Socket
+import           Text.Printf
+import           System.Environment (getArgs)
+import           System.Console.GetOpt
+import           System.IO (hFlush, hPutStr, stderr, stdout)
+
+import           Network.Mux.Bearer.Socket
+import           Network.Mux.Types
+import           Network.Mux.Timeout
+
+mainnetMagic :: Word32
+mainnetMagic = 764824073
+
+handshakeNum ::  MiniProtocolNum
+handshakeNum = MiniProtocolNum 0
+
+keepaliveNum :: MiniProtocolNum
+keepaliveNum = MiniProtocolNum 8
+
+data Flag = HostF String | PortF String | MagicF String | QuietF | JsonF deriving Show
+
+options :: [OptDescr Flag]
+options = [
+    Option "h" ["h"]     (ReqArg HostF "host")  "hostname/ip, e.g relay.iohk.example",
+    Option "m" ["magic"] (ReqArg MagicF "magic") ("magic, defaults to " ++ show mainnetMagic),
+    Option "j" ["json"]  (NoArg JsonF ) "json output flag",
+    Option "p" ["port"]  (ReqArg PortF "port") "portnumber, e.g 1234",
+    Option "q" ["quiet"] (NoArg QuietF ) "quiet flag, csv/json only output."
+  ]
+
+getHost :: [Flag] -> String
+getHost [] = error "Specify host/ip with '-h <hostname>'"
+getHost (HostF h:_) = h
+getHost (_:xs) = getHost xs
+
+getPort :: [Flag] -> String
+getPort [] = error "Specify port with '-p <port number>'"
+getPort (PortF p:_) = p
+getPort (_:xs) = getPort xs
+
+getQuiet :: [Flag] -> Bool
+getQuiet [] = False
+getQuiet (QuietF:_) = True
+getQuiet (_:xs) = getQuiet xs
+
+getMagic :: [Flag] -> Word32
+getMagic [] = mainnetMagic
+getMagic (MagicF m:_) = Prelude.read m
+getMagic (_:xs) = getMagic xs
+
+getJson :: [Flag] -> Bool
+getJson [] = False
+getJson (JsonF:_) = True
+getJson (_:xs) = getJson xs
+
+main :: IO ()
+main = do
+    args <- getArgs
+    let (actions, _, _ ) = getOpt RequireOrder options args
+        host = getHost actions
+        port = getPort actions
+        quiet = getQuiet actions
+        magic = getMagic actions
+        jsonF = getJson actions
```

Comment:
Another way is to use a record type with a monoid instance and `foldMap` over `getArgs` results. It might be a bit shorter and more elegant: it would need the record type definition, `Semigroup` and `Monoid` instances and `foldMap` accumulator function.

Reply:
Reworked to use `foldr`